File: tests/algorithms/x/Java/dynamic_programming/minimum_steps_to_one.java
public class Main {

    static int[] make_list(int len, int value) {
        int[] arr = ((int[])(new int[]{}));
        int i = 0;
        while (i < len) {
            arr = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(arr), java.util.stream.IntStream.of(value)).toArray()));
            i = i + 1;
        }
        return arr;
    }

    static int min_int(int a, int b) {
        if (a < b) {
            return a;
        }
        return b;
    }

    static int min_steps_to_one(int number) {
        if (number <= 0) {
            return 0;
        }
        int[] table = ((int[])(make_list(number + 1, number + 1)));
table[1] = 0;
        int i_1 = 1;
        while (i_1 < number) {
table[i_1 + 1] = min_int(table[i_1 + 1], table[i_1] + 1);
            if (i_1 * 2 <= number) {
table[i_1 * 2] = min_int(table[i_1 * 2], table[i_1] + 1);
            }
            if (i_1 * 3 <= number) {
table[i_1 * 3] = min_int(table[i_1 * 3], table[i_1] + 1);
            }
            i_1 = i_1 + 1;
        }
        return table[number];
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            System.out.println(_p(min_steps_to_one(10)));
            System.out.println(_p(min_steps_to_one(15)));
            System.out.println(_p(min_steps_to_one(6)));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/dynamic_programming/minimum_steps_to_one.pas
{$mode objfpc}
program Main;
uses SysUtils;
type IntArray = array of integer;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  len: integer;
  value: integer;
  b: integer;
  a: integer;
  number: integer;
function make_list(len: integer; value: integer): IntArray; forward;
function min_int(a: integer; b: integer): integer; forward;
function min_steps_to_one(number: integer): integer; forward;
function make_list(len: integer; value: integer): IntArray;
var
  make_list_arr: array of integer;
  make_list_i: integer;
begin
  make_list_arr := [];
  make_list_i := 0;
  while make_list_i < len do begin
  make_list_arr := concat(make_list_arr, IntArray([value]));
  make_list_i := make_list_i + 1;
end;
  exit(make_list_arr);
end;
function min_int(a: integer; b: integer): integer;
begin
  if a < b then begin
  exit(a);
end;
  exit(b);
end;
function min_steps_to_one(number: integer): integer;
var
  min_steps_to_one_table: IntArray;
  min_steps_to_one_i: integer;
begin
  if number <= 0 then begin
  exit(0);
end;
  min_steps_to_one_table := make_list(number + 1, number + 1);
  min_steps_to_one_table[1] := 0;
  min_steps_to_one_i := 1;
  while min_steps_to_one_i < number do begin
  min_steps_to_one_table[min_steps_to_one_i + 1] := min_int(min_steps_to_one_table[min_steps_to_one_i + 1], min_steps_to_one_table[min_steps_to_one_i] + 1);
  if (min_steps_to_one_i * 2) <= number then begin
  min_steps_to_one_table[min_steps_to_one_i * 2] := min_int(min_steps_to_one_table[min_steps_to_one_i * 2], min_steps_to_one_table[min_steps_to_one_i] + 1);
end;
  if (min_steps_to_one_i * 3) <= number then begin
  min_steps_to_one_table[min_steps_to_one_i * 3] := min_int(min_steps_to_one_table[min_steps_to_one_i * 3], min_steps_to_one_table[min_steps_to_one_i] + 1);
end;
  min_steps_to_one_i := min_steps_to_one_i + 1;
end;
  exit(min_steps_to_one_table[number]);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(IntToStr(min_steps_to_one(10)));
  writeln(IntToStr(min_steps_to_one(15)));
  writeln(IntToStr(min_steps_to_one(6)));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
